src/wmtk/operations/composite/TetEdgeSwap.cpp

```diff
@@ -0,0 +1,150 @@
+#include "TetEdgeSwap.hpp"
+#include <wmtk/simplex/closed_star.hpp>
+#include <wmtk/simplex/open_star.hpp>
+#include <wmtk/simplex/utils/SimplexComparisons.hpp>
+
+
+#include <wmtk/Mesh.hpp>
+
+namespace wmtk::operations::composite {
+TetEdgeSwap::TetEdgeSwap(Mesh& m, long collapse_index)
+    : Operation(m)
+    , m_split(m)
+    , m_collapse(m)
+    , m_collapse_index(collapse_index)
+{}
+
+std::vector<Simplex> TetEdgeSwap::execute(const Simplex& simplex)
+{
+    const auto split_simplicies = m_split(simplex);
+    if (split_simplicies.empty()) return {};
+    assert(split_simplicies.size() == 1);
+
+    // after split
+    const Tuple& split_ret = split_simplicies.front().tuple();
+    assert(!mesh().is_boundary(split_ret));
+    auto iter_tuple = split_ret;
+    std::vector<Tuple> candidate_edge_tuples;
+    while (true) {
+        candidate_edge_tuples.push_back(mesh().switch_edge(iter_tuple));
+        iter_tuple = mesh().switch_tetrahedron(mesh().switch_face(iter_tuple));
+        if (iter_tuple == split_ret) {
+            break;
+        }
+    }
+
+    assert(m_collapse_index < candidate_edge_tuples.size());
```

Comment:
i thought the point was that you would run multiple passes - wouldn't this make the whole system stop running after hitting some edge with a valence below your desired valence?